DT: new pinctrl helpers

mbolivar-nordic opened the pull request:
The commit which introduces new pin control helper macros to devicetree.h is:

```
devicetree: add devicetree/pinctrl.h

This contains accessor macros for getting phandles out of pinctrl
properties by name and index. As usual, the representation in C for a
phandle is a node identifier.

Add these new macros:

- DT_PINCTRL_BY_IDX(node_id, pc_idx, idx): phandle at index idx
  in the pinctrl-<pc_idx> property

- DT_PINCTRL_0(node_id, idx): pinctrl-0 convenience for the same

- DT_PINCTRL_BY_NAME(node_id, name, idx): phandle at index idx
  in the pinctrl property named 'name'

- DT_PINCTRL_NAME_TO_IDX(node_id, name): convert a pinctrl property
  name to its index number

- DT_NUM_PINCTRLS_BY_IDX(node_id, pc_idx): number of phandles in
  pinctrl-<pc_idx>

- DT_NUM_PINCTRLS_BY_NAME(node_id, name): number of phandles in a
  named pinctrl property

- DT_NUM_PINCTRL_STATES(node_id): total number of pinctrl-<pc_idx>
  properties

- DT_PINCTRL_HAS_IDX(node_id, pc_idx): does pinctrl-<pc_idx> exist?

- DT_PINCTRL_HAS_NAME(node_id, name): does a named pinctrl property
  exist?

- DT_PINCTRL_IDX_TO_NAME_TOKEN(node_id, pc_idx): convert a pinctrl
  index to its name as a token, similar to DT_STRING_TOKEN()

- DT_PINCTRL_IDX_TO_NAME_UPPER_TOKEN(node_id, pc_idx): like
  DT_PINCTRL_IDX_TO_NAME_TOKEN, but with an uppercase result

As well as DT_DRV_INST equivalents, which take inst wherever node_id
appears above:

- DT_INST_PINCTRL_BY_IDX()
- DT_INST_PINCTRL_0()
- DT_INST_PINCTRL_BY_NAME()
- DT_INST_PINCTRL_NAME_TO_IDX()
- DT_INST_NUM_PINCTRLS_BY_IDX()
- DT_INST_NUM_PINCTRLS_BY_NAME()
- DT_INST_NUM_PINCTRL_STATES()
- DT_INST_PINCTRL_HAS_IDX()
- DT_INST_PINCTRL_HAS_NAME()
- DT_INST_PINCTRL_IDX_TO_NAME_TOKEN()
- DT_INST_PINCTRL_IDX_TO_NAME_UPPER_TOKEN()


Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
```

The rest of the PR is prep work, bookkeeping, and documentation.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/37644)
<!-- Reviewable:end -->

mbolivar-nordic committed:
scripts: gen_defines: fix typo

Missing 's'.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: refactor write_special_props

Move the partition handling code into its own function and rework the
comment. This is prep work for adding additional generated macros to
this function.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
devicetree: add devicetree/pinctrl.h

This contains accessor macros for getting phandles out of pinctrl
properties by name and index. As usual, the representation in C for a
phandle is a node identifier.

Add these new macros:

- DT_PINCTRL_BY_IDX(node_id, pc_idx, idx): phandle at index idx
  in the pinctrl-<pc_idx> property

- DT_PINCTRL_0(node_id, idx): pinctrl-0 convenience for the same

- DT_PINCTRL_BY_NAME(node_id, name, idx): phandle at index idx
  in the pinctrl property named 'name'

- DT_PINCTRL_NAME_TO_IDX(node_id, name): convert a pinctrl property
  name to its index number

- DT_NUM_PINCTRLS_BY_IDX(node_id, pc_idx): number of phandles in
  pinctrl-<pc_idx>

- DT_NUM_PINCTRLS_BY_NAME(node_id, name): number of phandles in a
  named pinctrl property

- DT_NUM_PINCTRL_STATES(node_id): total number of pinctrl-<pc_idx>
  properties

- DT_PINCTRL_HAS_IDX(node_id, pc_idx): does pinctrl-<pc_idx> exist?

- DT_PINCTRL_HAS_NAME(node_id, name): does a named pinctrl property
  exist?

- DT_PINCTRL_IDX_TO_NAME_TOKEN(node_id, pc_idx): convert a pinctrl
  index to its name as a token, similar to DT_STRING_TOKEN()

- DT_PINCTRL_IDX_TO_NAME_UPPER_TOKEN(node_id, pc_idx): like
  DT_PINCTRL_IDX_TO_NAME_TOKEN, but with an uppercase result

As well as DT_DRV_INST equivalents, which take inst wherever node_id
appears above:

- DT_INST_PINCTRL_BY_IDX()
- DT_INST_PINCTRL_0()
- DT_INST_PINCTRL_BY_NAME()
- DT_INST_PINCTRL_NAME_TO_IDX()
- DT_INST_NUM_PINCTRLS_BY_IDX()
- DT_INST_NUM_PINCTRLS_BY_NAME()
- DT_INST_NUM_PINCTRL_STATES()
- DT_INST_PINCTRL_HAS_IDX()
- DT_INST_PINCTRL_HAS_NAME()
- DT_INST_PINCTRL_IDX_TO_NAME_TOKEN()
- DT_INST_PINCTRL_IDX_TO_NAME_UPPER_TOKEN()

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_defines: emit pinctrl helper macros

We need to be able to access pinctrl-<index> property contents by
name, convert names to indexes, convert indexes to names, and perform
existence checks by name and by index.

This is currently not possible because we don't track these properties
the same way we do other named properties. That in turn is because
they are different then the usual named properties.

The usual case looks like this, picking DMAs just for the sake of
example:

  dmas = <&dma0 ...>, <&dma1 ...>;
  dma-names = "tx", "rx";

So "tx" is the name for the <&dma0 ...> element, and "rx" is the name
for the <&dma1 ...> element, all in a single "dmas" property.

By contrast, pinctrl properties look like this:

  pinctrl-0 = <&foo &bar ...>;
  pinctrl-1 = <&baz &blub ...>;
  pinctrl-names = "default", "sleep";

Here, "default" is the name for the entire pinctrl-0 property.
Similarly, "sleep" is the name of the pinctrl-1 property. It's a
strange situation where the node itself is kind of a container for an
array of pin control properties, which Zephyr's bindings language
can't really capture and has some special case handling in edtlib.

This is easiest to handle with ad-hoc code. Add special case macros
for pinctrls.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
devicetree: add tests for devicetree/pinctrl.h macros

Test cases for new APIs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dts: test: fix vnd,adc-temp-sensor binding path

This isn't following the <compatible>.yaml naming convention; fix it.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: document devicetree/pinctrl.h macros

Add a new section in the API reference for the newly added
devicetree/pinctrl.h macros.

Amend macros.bnf in the guide to reflect the new generated macros for
getting at pinctrl information by name.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
soc: atmel: use new DT pinctrl accessors

Update to use the new APIs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
soc: cypress: use new DT pinctrl accessors

Update to use the new APIs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
soc: nuvoton_npcx: use new DT pinctrl accessors

Update to use the new APIs. Fix a typo while I'm here.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
soc: riscv-ite: use new DT pinctrl accessors

Update to use the new APIs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
pwm: it8xxx2: use new DT pinctrl accessors

Update to use the new APIs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

gmarull commented on include/devicetree/pinctrl.h:
```diff
@@ -0,0 +1,166 @@
+ * @return node identifier for the phandle at index 0 in the pinctrl-0
+ *         property of that node
+ */
+#define DT_PINCTRL_0(node_id) DT_PINCTRL_BY_IDX(node_id, 0, 0)
```
I'm not sure we should add this one. I think that `pinctrl` is better represented using states, so drivers should work using the state concept and not indices, so I'm not sure if providing extra helpers for indices helps on that front.

This code:
```
      n: node {
              pinctrl-0 = <&foo &bar>;
              pinctrl-1 = <&baz &blub>;
              pinctrl-names = "default", "sleep";
      }
 ```

and this other one:
```
      n: node {
              pinctrl-0 = <&foo &bar>;
              pinctrl-1 = <&baz &blub>;
              pinctrl-names = "sleep", "default";
      }
 ```

should result in same behavior, ie indices are not relevant.

mbolivar-nordic replied:
> so drivers should work using the state concept and not indices

This seems to be the main assertion here. The fact is that drivers are currently using indices, not just states. Now, that may be because there is no support for names yet. I don't know.

Does Linux use states exclusively, or does it also use indices?

gmarull replied:
Yes, Linux uses states: https://www.kernel.org/doc/html/latest/driver-api/pin-control.html#pin-control-requests-from-drivers

As far as I know Zephyr has never had such concept (probably because DT API/tooling was missing and/or current pinctrl drivers are not in a mature status).

galak replied:
I see this as no different than any other `name`d property in devicetree.  Its left to the binding and driver to decide if it wants to use the named version or not, and if the order of the values in an "array" has meaning.  For example regs and irqs do this all the time with assumptions in binding/driver.

gmarull replied:
Whether it's the driver or not who takes care of storing states, names makes things easier. If every driver decides to use indices on their own, bindings will need to specify "use index 0 for default state and 1 for sleep".

If I have to write a driver that needs to obtain a particular IRQ (e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/st/h7/stm32h7.dtsi#L452) I'll always obtain it by name, just to make sure I get what I want regardless of the order.

galak replied:
> Whether it's the driver or not who takes care of storing states, names makes things easier. If every driver decides to use indices on their own, bindings will need to specify "use index 0 for default state and 1 for sleep".
> 
> If I have to write a driver that needs to obtain a particular IRQ (e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/st/h7/stm32h7.dtsi#L452) I'll always obtain it by name, just to make sure I get what I want regardless of the order.

I agree that is better practice, but from an API point of view it makes sense to have what @mbolivar-nordic proposes in this PR.  How that gets used is left the the driver developers.

gmarull replied:
> I agree that is better practice, but from an API point of view it makes sense to have what @mbolivar-nordic proposes in this PR. How that gets used is left the the driver developers.

If we think pinctrl representation is better understood using states, as in Linux, then indices should have no relevance. We now use index 0 everywhere because it's really the only thing we have, but as soon as more states start to be implemented then `0` will have no special meaning other than being the first entry. Summarizing, I think we should be careful on adding new APIs, specially because current pinctrl state is not mature enough. It may be useful as a transition step to make current code less verbose, though.

mbolivar-nordic replied:
Trying to summarize:

- @gmarull wants Zephyr to use states exclusively, thinks using indices is a bad practice, points at Linux pinmux APIs using states
- @galak agrees using names is better practice, but thinks this should be left up to the user to decide and the API should be neutral

I don't know what the project is going to decide re: states vs. indices.

However, if we do decide on a state-based approach, then doing all the index-based pin control access via dedicated APIs instead of directly accessing the pinctrl-0 and pinctrl-1 properties as we are doing now will make it trivial to deprecate the index-based way of doing things. So even if the API is not forever, I think it is better to introduce and use it than let the current status quo stand.

@gmarull would you agree with that argument?

gmarull commented on dts/bindings/test/vnd,adc-temp-sensor.yaml:
```diff
@@ -21,3 +21,9 @@ properties:
     clocks:
       required: true
+
+    pinctrl-0:
+      type: phandles
+
+    pinctrl-1:
+      type: phandles
```
should we provide something like this: https://github.com/zephyrproject-rtos/zephyr/blob/ca4c7d200ee6f1ef77dbfe1ffef8be537af140da/dts/bindings/pinctrl/pinctrl-device.yaml ?

galak replied:
@mbolivar-nordic I don't know if you want to add `pinctrl-0` & `pinctrl-1`, and `pinctrl-names` to dts/bindings/base/base.yaml as part of this PR.

mbolivar-nordic replied:
I think I'll just include pinctrl-device.yaml into vnd,adc-temp-sensor for now. I think base.yaml is pretty big already and it's hard for users to know when a property in there is relevant to their device or not.

mbolivar-nordic replied:
@gmarull actually, that commit doesn't seem to be upstream yet...

gmarull replied:
It's part of #37572, feel free to take that file.

mbolivar-nordic replied:
> It's part of #37572, feel free to take that file.

Thanks but no thanks. Pulling that file in without updating all the in-tree bindings to include it feels wrong to me, and that's outside the scope of this PR.

gmarull reviewed (commented):
Some initial thoughts, thanks for this DT update

galak reviewed (changes requested):
Looks good, one comment on what `_BY_IDX` means and impact on `PINCTRL_0` api variants.

mbolivar-nordic committed:
adc: npcx: use new DT pinctrl accessors

Update to use the new APIs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
dt-bindings: b91-pinctrl: use new DT pinctrl accessors

Update to use the new APIs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
soc: stm32: use new DT pinctrl accessors

Update to use the new APIs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@galak @gmarull please revisit. I updated the PR to hopefully address all the use cases and comments I could. I replied to comments individually when I didn't make any change.

ChiHuaL commented on soc/arm/nuvoton_npcx/common/soc_dt.h:
```diff
@@ -181,10 +162,10 @@
-		     NPCX_DT_ALT_ITEMS_FUNC,       \
-		     inst)                         \
+#define NPCX_DT_ALT_ITEMS_LIST(inst) {			\
+	UTIL_LISTIFY(DT_NUM_PINCTRLS_BY_IDX(inst, 0),	\
```
Because the parameter is "inst", it should use DT_DRV_INST equivalent macro here.
`
UTIL_LISTIFY(DT_INST_NUM_PINCTRLS_BY_IDX(inst, 0),      \
`

mbolivar-nordic replied:
Thanks @ChiHuaL; now fixed.

galak commented on include/devicetree/pinctrl.h:
```diff
@@ -0,0 +1,369 @@
+ * @param pc_idx index of the pinctrl property itself
+ * @return number of phandles in the property with that index
+ */
+#define DT_NUM_PINCTRLS_BY_IDX(node_id, pc_idx) \
```
Why is this not just `DT_PROP_LEN` of the property?  Not sure we need this macro.

mbolivar-nordic replied:
> Why is this not just `DT_PROP_LEN` of the property? Not sure we need this macro.

I explained a bit here:

https://github.com/zephyrproject-rtos/zephyr/pull/37621#discussion_r687987955

That's what people are doing now, but they are pasting ``pinctrl_##num`` in their code to do that. Pasting in cpp has "gotchas" that the DT_CATx macros we use under the hood make sure users don't run into.

For example if you do

```c
#define MY_INDEX 1
#define FOO(node_id, pc_idx) DT_PROP_LEN(node_id, pinctrl_ ## pc_idx)

void foo(void)
{
	printk("raw len with pasting: %d\n", FOO(TEST_TEMP, MY_INDEX));
	printk("len with api: %d\n", DT_NUM_PINCTRLS_BY_IDX(TEST_TEMP, MY_INDEX));
}
```

Then this expands to:

```c
void foo(void)
{
 printk("raw len with pasting: %d\n", DT_N_S_test_S_temperature_sensor_P_pinctrl_MY_INDEX_LEN);
 printk("len with api: %d\n", 2);
}
```

Because manual use of `##` in `FOO` misses that you actually have to recursively expand again to get `MY_INDEX` to resolve to an integer literal.

gmarull replied:
question: `DT_NUM_*` macros do not follow the `DT_PINCTRL_*` namespace convention, not sure if important.

mbolivar-nordic replied:
> question: `DT_NUM_*` macros do not follow the `DT_PINCTRL_*` namespace convention, not sure if important.

https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_NUM_REGS
https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_NUM_IRQS

Yep, done on purpose for consistency with other APIs like the above.

galak commented on include/devicetree/pinctrl.h:
```diff
@@ -0,0 +1,369 @@
+ * @return number of phandles in the property with that name
+ */
+#define DT_NUM_PINCTRLS_BY_NAME(node_id, name) \
+	DT_NUM_PINCTRLS_BY_IDX(node_id, DT_PINCTRL_TO_IDX(node_id, name))
```
same comment as above.

mbolivar-nordic replied:
> same comment as above.

resolving since it's the same response

galak reviewed (changes requested):
looks good, just question about need of DT_NUM_PINCTRLS_BY_IDX and variants.

mbolivar-nordic commented:
Last push is just documentation fixes.

mbolivar-nordic commented:
I'm adding a DNM while we're discussing the layers that will be built on top of this.

There are currently two draft proposals:

- https://github.com/zephyrproject-rtos/zephyr/pull/37621 (by @galak)
- https://github.com/zephyrproject-rtos/zephyr/pull/37572 (by @gmarull)

In either one of these, some of the macros introduced here and used in soc-specific headers may become redundant.

mbolivar-nordic committed:
edtlib: emit pinctrl names as tokens

This will be necessary to going back and forth between indexes and
names in C.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
Latest changes:

- add new macros (requested by @gmarull):
-- DT_PINCTRL_IDX_TO_NAME_TOKEN(node_id, pc_idx)
-- DT_PINCTRL_IDX_TO_NAME_UPPER_TOKEN(node_id, pc_idx)
-- DT_PINCTRL_IDX_TO_NAME_TOKEN()
-- DT_PINCTRL_IDX_TO_NAME_UPPER_TOKEN()
- make existing macros consume lowercase-and-underscores names tokens
- update tests
- add missing tests

brockus-zephyr reviewed (approved):
> Reviewed 2 of 3 files at r1, 18 of 18 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/37644)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @ABOSTM, @FRASTM, @henrikbrixandersen, @ioannisg, @jackrosenthal, @katsuster, @keith-zephyr, @kgugala, @MaureenHelm, @mgielda, @nandojve, @nashif, @pgielda, and @sjg20)



<!-- Sent from Reviewable.io -->

mbolivar-nordic commented:
Rebased and removed DNM. I'm going to leave it up to @galak to decide if we want this in for LTS>

cfriedt commented:
@galak - feel free to merge or if you would like to hold off apply the DNM label again

brockus-zephyr reviewed (approved):
> Reviewed 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/37644)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @ABOSTM, @FRASTM, @henrikbrixandersen, @ioannisg, @jackrosenthal, @katsuster, @keith-zephyr, @kgugala, @MaureenHelm, @mgielda, @nashif, @pgielda, and @sjg20)



<!-- Sent from Reviewable.io -->
